README.md
# Flutter Yandex SmartCaptcha

## Example:

```Dart
import 'package:flutter/material.dart';
import 'package:flutter_yandex_smartcaptcha/flutter_yandex_smartcaptcha.dart';

// Get you key from admin panel yandex cloud
String siteKey = const String.fromEnvironment('SITE_KEY');

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Yandex SmartCaptcha',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Example'),
    );
  }
}

/// CaptchaConfig:
/// sitekey — ключ клиентской части.
/// languageCaptcha — язык виджета. 'ru' | 'en' | 'be' | 'kk' | 'tt' | 'uk' | 'uz' | 'tr';
/// testMode — включение работы капчи в режиме тестирования. Пользователь всегда будет получать задание. Используйте это свойство только для отладки и тестирования.
/// isWebView — запуск капчи в WebView. Используется для повышения точности оценки пользователей при добавлении капчи в мобильные приложения с помощью WebView.
/// invisible — невидимая капча.
/// hideShield — скрыть блок с уведомлением об обработке данных.
/// colorBackground — цвет фона

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  late final CaptchaConfig captchaConfig;
  final Controller _controller = Controller();

  @override
  void initState() {
    super.initState();
    captchaConfig = CaptchaConfig(
        siteKey: siteKey,
        testMode: true,
        languageCaptcha: 'ru',
        invisible: false,
        isWebView: true,
        colorBackground: Colors.green
    );
    _controller.onReadyCallback(() {
      debugPrint('SmartCaptcha controller is ready');
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: Column(
          children: [
            Expanded(
                child: YandexSmartCaptcha(
                  captchaConfig: captchaConfig,
                  onLoadWidget: const Center(
                    child: Center(
                      child: SizedBox(height: 60, width: 60, child: CircularProgressIndicator()),
                    ),
                  ),
                  controller: _controller,
                  challengeViewCloseCallback: () {
                    debugPrint('call: challengeViewCloseCallback');
                  },
                  challengeViewOpenCallback: () {
                    debugPrint('call: challengeViewOpenCallback');
                  },
                  networkErrorCallback: () {
                    debugPrint('call: networkErrorCallback');
                  },
                  tokenResultCallback: (String token) {
                    debugPrint('call: tokenResultCallback $token');
                  },
                  shouldOpenPolicy: (String urlPolicy) {
                    if (urlPolicy.contains('smartcaptcha_notice')) {
                      return false;
                    } else {
                      return true;
                    }
                  },
                )),
            Padding(
              padding: const EdgeInsets.all(8.0),
              child: Row(
                mainAxisAlignment: MainAxisAlignment.spaceAround,
                children: [
                  ElevatedButton(
                      onPressed: () async {
                        final bool isReady = _controller.isControllerIsReady();
                        if (isReady) {
                          _controller.execute();
                        }
                      },
                      child: const Text('Execute')),
                  ElevatedButton(
                      onPressed: () async {
                        final bool isReady = _controller.isControllerIsReady();
                        if (isReady) {
                          _controller.destroy();
                        }
                      },
                      child: const Text('Destroy'))
                ],
              ),
            )
          ],
        ),
      ),
    );
  }
}

```

![image](https://github.com/postflow/flutterYandexSmartCaptcha/blob/master/pic/android_small.png?raw=true)
![image](https://github.com/postflow/flutterYandexSmartCaptcha/blob/master/pic/ios_small.png?raw=true)


CHANGELOG.md
## 0.0.1
* Add CaptchaConfig.

## 0.0.2
* Add Controller.

## 0.0.3
* Update example.

## 0.0.4
* Update readme.
## 0.0.5
* Update readme.

test/flutter_yandex_smartcaptcha_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_yandex_smartcaptcha/flutter_yandex_smartcaptcha.dart';

void main() {
  test('adds one to input values', () {

  });
}


lib/src/web_page_content.dart
//https://cloud.yandex.ru/docs/smartcaptcha/concepts/widget-methods#methods

final class WebPageCaptchaContent {
  final String _siteKey;
  final bool _testMode;
  final String _languageCaptcha;
  final bool _invisible;
  final bool _hideShield;
  final bool _isWebView;

  late String yandexSmartCaptchaWebContent;

  WebPageCaptchaContent({
    required String siteKey,
    bool testMode = false,
    String languageCaptcha = 'ru',
    bool invisible = false,
    bool hideShield = false,
    bool isWebView = true
  }) : _isWebView = isWebView, _hideShield = hideShield, _invisible = invisible, _languageCaptcha = languageCaptcha, _testMode = testMode, _siteKey = siteKey {

    yandexSmartCaptchaWebContent = '''
            <!doctype html>
            <html lang="ru">
            <head>
              <meta charset="utf-8" />
              <meta name="viewport" content="width=device-width, initial-scale=1">
              <title></title>
               <script src="https://smartcaptcha.yandexcloud.net/captcha.js?render=onload&onload=onLoadFunction" defer></script>
            </head>
            <body>
            <script>
              function onLoadFunction() {
                   if (window.smartCaptcha) {
                     const widgetId = window.smartCaptcha.render('captcha-container', {
                          sitekey: '$_siteKey',
                          invisible: $_invisible,
                          hideShield: $_hideShield,
                          hl: '$_languageCaptcha',
                          test: $_testMode,
                          webview: $_isWebView,
                          callback: resultCallback
                      });
              
                      window.smartCaptcha.subscribe(
                          widgetId,
                          'challenge-visible',
                          () => { window.flutter_inappwebview.callHandler('challengeVisibleEvent'); }
                      );
              
                      window.smartCaptcha.subscribe(
                          widgetId,
                          'challenge-hidden',
                          () => { window.flutter_inappwebview.callHandler('challengeHiddenEvent'); }
                      );
                      
                  
                      window.smartCaptcha.subscribe(
                          widgetId,
                          'network-error',
                          () => { window.flutter_inappwebview.callHandler('networkErrorEvent'); }
                      );
                      
                       window.flutter_inappwebview.callHandler('captchaContainerIsLoaded');
                      
                    
                  } else {
                    window.flutter_inappwebview.callHandler('networkErrorEvent');
                  }
              }
              
              function resultCallback(token) {
                  window.flutter_inappwebview.callHandler('tokenHandler', token);
              }
              
            </script>
                <div id="captcha-container" style="height: 100px"></div>
            </body>
            </html>
    ''';
  }
}


lib/flutter_yandex_smartcaptcha.dart
library flutter_yandex_smartcaptcha;

import 'package:flutter/material.dart';
import 'package:flutter_inappwebview/flutter_inappwebview.dart';
import 'package:flutter_yandex_smartcaptcha/src/onload_placeholder_widget.dart';
import 'package:flutter_yandex_smartcaptcha/src/web_page_content.dart';

export 'package:flutter_yandex_smartcaptcha/flutter_yandex_smartcaptcha.dart';

/// sitekey — ключ клиентской части.
/// languageCaptcha — язык виджета. 'ru' | 'en' | 'be' | 'kk' | 'tt' | 'uk' | 'uz' | 'tr';
/// testMode — включение работы капчи в режиме тестирования. Пользователь всегда будет получать задание. Используйте это свойство только для отладки и тестирования.
/// isWebView — запуск капчи в WebView. Используется для повышения точности оценки пользователей при добавлении капчи в мобильные приложения с помощью WebView.
/// invisible — невидимая капча.
/// hideShield — скрыть блок с уведомлением об обработке данных.
/// colorBackground — цвет фона

class CaptchaConfig {
  String siteKey;
  bool testMode;
  String languageCaptcha;
  bool invisible;
  bool hideShield;
  bool isWebView;
  Color? colorBackground;

  CaptchaConfig(
      {required this.siteKey,
      this.testMode = false,
      this.languageCaptcha = 'ru',
      this.invisible = false,
      this.hideShield = false,
      this.isWebView = true,
      this.colorBackground});
}

class Controller {
  InAppWebViewController? _inAppWebViewController;
  VoidCallback? _readyCallback;

  Future destroy() async {
    return _inAppWebViewController?.evaluateJavascript(source: "window.smartCaptcha.destroy()");
  }

  Future execute() async {
    return _inAppWebViewController?.evaluateJavascript(source: "window.smartCaptcha.execute()");
  }

  void _setController(InAppWebViewController controller) {
    _inAppWebViewController = controller;
    _readyCallback?.call();
  }

  bool isControllerIsReady() => _inAppWebViewController != null ? true : false;

  onReadyCallback(VoidCallback readyCallback) {
    _readyCallback = readyCallback;
  }

}

class YandexSmartCaptcha extends StatefulWidget {
  final CaptchaConfig captchaConfig;
  final Controller? controller;
  final VoidCallback? challengeViewOpenCallback;
  final VoidCallback? challengeViewCloseCallback;
  final VoidCallback? networkErrorCallback;
  final Function(String) tokenResultCallback;
  final bool Function(String urlPolicy)? shouldOpenPolicy;
  final Widget? onLoadWidget;

  const YandexSmartCaptcha({
      required this.captchaConfig,
      this.controller,
      required this.tokenResultCallback,
      this.challengeViewCloseCallback,
      this.challengeViewOpenCallback,
      this.networkErrorCallback,
      this.shouldOpenPolicy,
      this.onLoadWidget,
      Key? key}) : super(key: key);

  @override
  State<YandexSmartCaptcha> createState() => _YandexSmartCaptchaState();
}

class _YandexSmartCaptchaState extends State<YandexSmartCaptcha>  {
  late final WebPageCaptchaContent _webPageCaptchaContent;

  Controller? _controller;
  bool _webViewIsCreated = false;

  final InAppWebViewGroupOptions options = InAppWebViewGroupOptions(
      crossPlatform: InAppWebViewOptions(
        transparentBackground: true,
        useShouldOverrideUrlLoading: true,
        mediaPlaybackRequiresUserGesture: false,
      ),
      android: AndroidInAppWebViewOptions(
        useHybridComposition: true,
      ),
      ios: IOSInAppWebViewOptions(
        allowsInlineMediaPlayback: true,
      ));

  @override
  void initState() {
    super.initState();
    _controller = widget.controller;
    final CaptchaConfig captchaConfig = widget.captchaConfig;
    _webPageCaptchaContent = WebPageCaptchaContent(
        siteKey: captchaConfig.siteKey,
        testMode: captchaConfig.testMode,
        languageCaptcha: captchaConfig.languageCaptcha,
        invisible: captchaConfig.invisible,
        hideShield: captchaConfig.hideShield,
        isWebView: captchaConfig.isWebView
    );
  }

  @override
  Widget build(BuildContext context) {
    return Stack(
        children: [
          if (widget.captchaConfig.colorBackground != null)  Container(color: widget.captchaConfig.colorBackground),
          if (widget.onLoadWidget != null) ...[OnLoadPlaceholder(onLoadWidget: widget.onLoadWidget, webViewIsCreated: _webViewIsCreated)] ,
          InAppWebView(
            initialOptions: options,
            initialData: InAppWebViewInitialData(data: _webPageCaptchaContent.yandexSmartCaptchaWebContent),
            androidOnPermissionRequest: (controller, origin, resources) async {
              return PermissionRequestResponse(resources: resources, action: PermissionRequestResponseAction.GRANT);
            },
            shouldOverrideUrlLoading: (controller, shouldOverrideUrlLoadingRequest) async {
              final String url = shouldOverrideUrlLoadingRequest.request.url.toString();
              final bool callbackResult = widget.shouldOpenPolicy?.call(url) ?? true;
              return callbackResult ? Future.value(NavigationActionPolicy.ALLOW) : Future.value(NavigationActionPolicy.CANCEL);
            },
            onConsoleMessage: (controller, consoleMessage) {
              debugPrint('YandexSmartCaptcha js console message:${consoleMessage.toString()}');
            },
            onWebViewCreated: (InAppWebViewController controller) {
              _controller?._setController(controller);
              controller.addJavaScriptHandler(
                  handlerName: 'tokenHandler',
                  callback: (args) {
                    widget.tokenResultCallback.call(args.toString());
                  });

              controller.addJavaScriptHandler(
                  handlerName: 'challengeVisibleEvent',
                  callback: (args) {
                    widget.challengeViewOpenCallback?.call();
                  });

              controller.addJavaScriptHandler(
                  handlerName: 'challengeHiddenEvent',
                  callback: (args) {
                    widget.challengeViewCloseCallback?.call();
                  });

              controller.addJavaScriptHandler(
                  handlerName: 'networkErrorEvent',
                  callback: (args) {
                    widget.networkErrorCallback?.call();
                  });

              controller.addJavaScriptHandler(
                  handlerName: 'captchaContainerIsLoaded',
                  callback: (args) {
                    setState(() {
                      _webViewIsCreated = true;
                    });
                  });
            },
          )
        ]);
  }
}

example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_yandex_smartcaptcha/flutter_yandex_smartcaptcha.dart';

// Get you key from admin panel yandex cloud
String siteKey = const String.fromEnvironment('SITE_KEY');

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Yandex SmartCaptcha',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Example'),
    );
  }
}

/// CaptchaConfig:
/// sitekey — ключ клиентской части.
/// languageCaptcha — язык виджета. 'ru' | 'en' | 'be' | 'kk' | 'tt' | 'uk' | 'uz' | 'tr';
/// testMode — включение работы капчи в режиме тестирования. Пользователь всегда будет получать задание. Используйте это свойство только для отладки и тестирования.
/// isWebView — запуск капчи в WebView. Используется для повышения точности оценки пользователей при добавлении капчи в мобильные приложения с помощью WebView.
/// invisible — невидимая капча.
/// hideShield — скрыть блок с уведомлением об обработке данных.
/// colorBackground — цвет фона

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  late final CaptchaConfig captchaConfig;
  final Controller _controller = Controller();

  @override
  void initState() {
    super.initState();
    captchaConfig = CaptchaConfig(
        siteKey: siteKey,
        testMode: true,
        languageCaptcha: 'ru',
        invisible: false,
        isWebView: true,
        colorBackground: Colors.green
    );
    _controller.onReadyCallback(() {
      debugPrint('SmartCaptcha controller is ready');
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: Column(
          children: [
            Expanded(
              child: YandexSmartCaptcha(
              captchaConfig: captchaConfig,
              onLoadWidget: const Center(
                child: Center(
                  child: SizedBox(height: 60, width: 60, child: CircularProgressIndicator()),
                ),
              ),
              controller: _controller,
              challengeViewCloseCallback: () {
                debugPrint('call: challengeViewCloseCallback');
              },
              challengeViewOpenCallback: () {
                debugPrint('call: challengeViewOpenCallback');
              },
              networkErrorCallback: () {
                debugPrint('call: networkErrorCallback');
              },
              tokenResultCallback: (String token) {
                debugPrint('call: tokenResultCallback $token');
              },
              shouldOpenPolicy: (String urlPolicy) {
                if (urlPolicy.contains('smartcaptcha_notice')) {
                  return false;
                } else {
                  return true;
                }
              },
            )),
            Padding(
              padding: const EdgeInsets.all(8.0),
              child: Row(
                mainAxisAlignment: MainAxisAlignment.spaceAround,
                children: [
                  ElevatedButton(
                      onPressed: () async {
                        final bool isReady = _controller.isControllerIsReady();
                        if (isReady) {
                          _controller.execute();
                        }
                      },
                      child: const Text('Execute')),
                  ElevatedButton(
                      onPressed: () async {
                        final bool isReady = _controller.isControllerIsReady();
                        if (isReady) {
                          _controller.destroy();
                        }
                      },
                      child: const Text('Destroy'))
                ],
              ),
            )
          ],
        ),
      ),
    );
  }
}
